drivers/platform/stm32/stm32_i2c.c

```diff
@@ -0,0 +1,188 @@
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include <errno.h>
+#include "util.h"
+#include "i2c.h"
+#include "stm32_i2c.h"
+
+/**
+ * @brief Initialize the I2C communication peripheral.
+ * @param desc - The I2C descriptor.
+ * @param param - The structure that contains the I2C parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t stm32_i2c_init(struct i2c_desc **desc,
+		       const struct i2c_init_param *param)
+{
+	int32_t ret;
+	struct i2c_desc	*i2c_desc;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	i2c_desc = (struct i2c_desc *)calloc(1, sizeof(*i2c_desc));
+	if (!i2c_desc)
+		return -ENOMEM;
+
+	struct stm32_i2c_desc *xdesc;
+	struct stm32_i2c_init_param *xinit;
+
+	xdesc = (stm32_i2c_desc*)calloc(1,sizeof(stm32_i2c_desc));
+	if (!xdesc) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	i2c_desc->extra = xdesc;
+	xinit = param->extra;
+
+	xdesc->hi2c.Instance = xinit->base;
+	xdesc->hi2c.Init.ClockSpeed = param->max_speed_hz;
+	xdesc->hi2c.Init.DutyCycle = I2C_DUTYCYCLE_2;
+	xdesc->hi2c.Init.OwnAddress1 = 0;
+	xdesc->hi2c.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+	xdesc->hi2c.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+	xdesc->hi2c.Init.OwnAddress2 = 0;
+	xdesc->hi2c.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
+	xdesc->hi2c.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+
+	ret = HAL_I2C_Init(&xdesc->hi2c);
+	if (ret != HAL_OK) {
+		ret = -EIO;
+		goto error;
+	}
+
+	/* copy settings to device descriptor */
+	i2c_desc->max_speed_hz = param->max_speed_hz;
+	i2c_desc->slave_address = param->slave_address;
+	*desc = i2c_desc;
+
+	return SUCCESS;
+error:
+	free(i2c_desc);
+	free(xdesc);
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by i2c_init().
+ * @param desc - The I2C descriptor.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t stm32_i2c_remove(struct i2c_desc *desc)
+{
+	struct stm32_i2c_desc *sdesc;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	sdesc = desc->extra;
+	HAL_I2C_DeInit(&sdesc->hi2c);
+	free(desc->extra);
+	free(desc);
+	return 0;
```

Comment:
nit: let's use `error.h` defines.

Reply:
i think i'll leave it as-is because we will abolish SUCCESS/FAILURE at some point anyway